lib/mayaUsd/ufe/UsdShaderNodeDef.cpp

```diff
@@ -81,12 +100,211 @@ UsdShaderNodeDef::UsdShaderNodeDef(const PXR_NS::SdrShaderNodeConstPtr& shaderNo
+{
+    if (fShaderNodeDef) {
+        if (PXR_NS::SdrShaderPropertyConstPtr property
+            = fShaderNodeDef->GetShaderInput(PXR_NS::TfToken(name))) {
+            return Ufe::AttributeDef::ConstPtr(new UsdShaderAttributeDef(property));
+        }
+    }
+    return {};
+}
+
+Ufe::ConstAttributeDefs UsdShaderNodeDef::inputs() const
+{
+    return fShaderNodeDef ? getAttrs<Ufe::AttributeDef::INPUT_ATTR>(fShaderNodeDef)
+                          : Ufe::ConstAttributeDefs();
+}
+
+std::vector<std::string> UsdShaderNodeDef::outputNames() const
+{
+    std::vector<std::string> retVal;
+    if (fShaderNodeDef) {
+        for (auto&& n : fShaderNodeDef->GetOutputNames()) {
+            retVal.push_back(n.GetString());
+        }
+    }
+    return retVal;
+}
+
+bool UsdShaderNodeDef::hasOutput(const std::string& name) const
+{
+    return fShaderNodeDef ? fShaderNodeDef->GetShaderOutput(PXR_NS::TfToken(name)) : false;
+}
+
+Ufe::AttributeDef::ConstPtr UsdShaderNodeDef::output(const std::string& name) const
+{
+    if (fShaderNodeDef) {
+        if (PXR_NS::SdrShaderPropertyConstPtr property
+            = fShaderNodeDef->GetShaderOutput(PXR_NS::TfToken(name))) {
+            return Ufe::AttributeDef::ConstPtr(new UsdShaderAttributeDef(property));
+        }
+    }
+    return {};
+}
+
+Ufe::ConstAttributeDefs UsdShaderNodeDef::outputs() const
+{
+    return fShaderNodeDef ? getAttrs<Ufe::AttributeDef::OUTPUT_ATTR>(fShaderNodeDef)
+                          : Ufe::ConstAttributeDefs();
+}
+
+Ufe::Value UsdShaderNodeDef::getMetadata(const std::string& key) const
+{
+    if (fShaderNodeDef) {
+        const PXR_NS::NdrTokenMap& metadata = fShaderNodeDef->GetMetadata();
+        auto it = metadata.find(PXR_NS::TfToken(key));
+        if (it != metadata.cend()) {
+            return Ufe::Value(it->second);
+        }
+    }
+    // TODO: Adapt UI metadata information found in SdrShaderNode to Ufe standards
+    // TODO: Fix Mtlx parser in USD to populate UI metadata in SdrShaderNode
+    return {};
+}
+
+bool UsdShaderNodeDef::hasMetadata(const std::string& key) const
+{
+    if (fShaderNodeDef) {
+        const PXR_NS::NdrTokenMap& metadata = fShaderNodeDef->GetMetadata();
+        auto it = metadata.find(PXR_NS::TfToken(key));
+        if (it != metadata.cend()) {
+            return true;
+        }
+    }
+    return false;
+}
+#endif
+
+#if (UFE_PREVIEW_VERSION_NUM >= 4010)
+Ufe::SceneItem::Ptr UsdShaderNodeDef::createNode(
+    const Ufe::SceneItem::Ptr& parent,
+    const Ufe::PathComponent&  name) const
+{
+    Ufe::SceneItem::Ptr createdItem = nullptr;
+
+    if (fShaderNodeDef) {
+        UsdSceneItem::Ptr parentItem = std::dynamic_pointer_cast<UsdSceneItem>(parent);
+        UsdUndoCreateFromNodeDefCommand::Ptr cmd
+            = UsdUndoCreateFromNodeDefCommand::create(fShaderNodeDef, parentItem, name.string());
+        if (cmd) {
+            cmd->execute();
```

Comment:
I know we do this a lot, mostly as a convenience to avoid having to code command and non-command versions of the UFE API, but it does feel awkward that we can't directly call what UsdUndoCreateFromNodeDefCommand is calling, and have to create a command object and call execute() on it.  Couldn't there be some kind of static engine method on UsdUndoCreateFromNodeDefCommand that would do the work without a command object?

Reply:
How do we handle all those commands that have validation in the constructor? The create node command ends up using UsdUndoAddNewPrimCommand and will properly react if the prim can not be added.